controllers/leaderelection/agent_leaderelection.go

```diff
@@ -0,0 +1,181 @@
+/*
+ * (c) Copyright IBM Corp. 2021
+ * (c) Copyright Instana Inc. 2021
+ */
+
+package leaderelection
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/instana/instana-agent-operator/controllers/leaderelection/coordination_api"
+	"github.com/procyon-projects/chrono"
+	coreV1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	LeaderElectionTask          chrono.ScheduledTask
+	LeaderElectionTaskScheduler chrono.TaskScheduler
+	KubernetesLeaderResourceId  = "com.instana.plugin.kubernetes.leader"
+
+	coordinationApi = coordination_api.New()
+)
+
+type LeaderElector struct {
+	Ctx    context.Context
+	Client client.Client
+	Scheme *runtime.Scheme
+	Log    logr.Logger
+}
+
+func (l *LeaderElector) StartCoordination(agentNameSpace string) error {
+	l.Log = ctrl.Log.WithName("leaderelector").WithName("InstanaAgent")
+	LeaderElectionTaskScheduler = chrono.NewDefaultTaskScheduler()
+	var err error
+	LeaderElectionTask, err = LeaderElectionTaskScheduler.ScheduleWithFixedDelay(func(ctx context.Context) {
+		var activePods map[string]coreV1.Pod
+		var err error
+		if activePods, err = l.fetchPods(agentNameSpace); err != nil {
+			l.Log.Error(err, "Unable to fetch agent pods for doing election")
+			return
+		}
+		l.pollAgentsAndAssignLeaders(activePods)
+
+		time.Sleep(5 * time.Second)
+	}, 5*time.Second)
+	if err == nil {
+		l.Log.Info("Task has been scheduled successfully.")
+	}
+	return err
+}
+
+func (l *LeaderElector) CancelLeaderElection() {
+	if LeaderElectionTask != nil && !LeaderElectionTask.IsCancelled() {
+		LeaderElectionTask.Cancel()
+	}
+	LeaderElectionTaskScheduler.Shutdown()
+}
+
+func (l *LeaderElector) fetchPods(agentNameSpace string) (map[string]coreV1.Pod, error) {
+	podList := &coreV1.PodList{}
+	activePods := make(map[string]coreV1.Pod)
+	lbs := map[string]string{
+		"app.kubernetes.io/name": "instana-agent",
+	}
+	labelSelector := labels.SelectorFromSet(lbs)
+	listOps := &client.ListOptions{Namespace: agentNameSpace, LabelSelector: labelSelector}
+	if err := l.Client.List(l.Ctx, podList, listOps); err != nil {
+		return nil, err
+	}
+
+	for _, pod := range podList.Items {
+		if pod.Status.Phase == coreV1.PodRunning {
+			activePods[string(pod.GetObjectMeta().GetUID())] = pod
+		}
+	}
+	return activePods, nil
+}
+
+func (l *LeaderElector) pollAgentsAndAssignLeaders(pods map[string]coreV1.Pod) {
+	for {
+		leadershipStatus := l.pollLeadershipStatus(pods)
+
+		if len(leadershipStatus.Status) == 0 {
```

Comment:
Question, what does this mean, what are you checking exactly? (I see its an empty list, but what does that mean? comment might help here)

Reply:
if leadershipStatus.Status is empty list that means we were not able to get leadershipstatus for any pod so we can't proceed with leadership assigning process and we should return, I added that now as a comment